smolyay/basis.py

```diff
@@ -0,0 +1,226 @@
+import abc
+import math
+
+import numpy
+
+class BasisFunction(abc.ABC):
+    """Abstract class for basis functions.
+
+    Creates a framework for basis functions that are the building blocks
+    for the terms in the surrogate model. 
+    ``points`` property represents all the unique 1D points of the 
+    basis function associated with a Smolyak index described by the 
+    class IndexGrid.
+
+    """
+
+    def __init__(self):
+        self._points = []
+
+    @abc.abstractproperty
+    def points(self):
+        """The points of the basis function assigned to Smolyak indices"""
+        pass
```

Comment:
The Chebyshev polynomial of the 1st kind can use both its extrema in addition to its root because its extrema within the domain [-1,1] lie in the range [-1,1]. The Chebyshev polynomial of the 2nd kind's extrema for that domain don't lie in that range which makes them not very useful. Having both be useful seems pretty unique and I don't think that distinction makes a lot of sense for many basis function besides Chebyshev of the 1st kind.

Reply:
OK, that's fair. We can just be clear in the documentation then, once you add the ability to choose how `points` is defined.